Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+        private bool rotateAccordingToSurface = false;
+
+        /// <summary>
+        /// If false, the game object to place will not change its rotation according to the surface hit.  The object will 
+        /// remain facing the camera while it is in the placing state.  If true, the object will rotate according to the surface normal
+        /// if there is a hit."
+        /// </summary>
+        public bool RotateAccordingToSurface
+        {
+            get => rotateAccordingToSurface;
+            set
+            {
+                if (rotateAccordingToSurface != value)
+                {
+                    rotateAccordingToSurface = value;
+                }
+            }
+        }
+
+        private const int IgnoreRaycastLayer = 2;
+
+        private const int DefaultLayer = 0;
+
+        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint)
+        protected RayStep currentRay;
+
+        protected RaycastHit currentHit;
+
+        protected float previousFrameNumber = 0;
+
+        protected override void Start()
+        {
+            // Solver base class
+            base.Start();
+
+            // If tap to place is added via script, set the GameObjectToPlace as this gameobject 
+            if (GameObjectToPlace == null)
+            {
+                GameObjectToPlace = gameObject;
+            }
+
+            if(!ColliderPresent)
+            {
+                Debug.LogError("The GameObjectToPlace does not have a collider attached, please attach a collider");
+            }
+
+            SurfaceNormalOffset = GetComponent<Collider>().bounds.extents.z;
+
+            // What if the spatial awareness is not enabled?
+            spatialAwarenessSystemAcecess = CoreServices.SpatialAwarenessSystem as IMixedRealityDataProviderAccess;
+
+            // Get the spatial mesh display option on start
+            // I do not know if this is legal, check on this
+            initialSpatialMeshDisplayOption = InitialSpatialMeshDisplayOption;
+
+            if (AutoStart)
+            {
+                StartPlacement();
+            }
+            else
+            {
+                SolverHandler.UpdateSolvers = false;
+            }
+        }
+
+        private void StartPlacement()
+        {
+            CoreServices.InputSystem?.RegisterHandler<IMixedRealityPointerHandler>(this);
+
+            // Temporarily change the game object layer to IgnoreRaycastLayer to enable a surface hit beyond the game object
+            gameObject.layer = IgnoreRaycastLayer;
+
+            if (SpatialMeshVisible)
+            {
+                ToggleSpatialMeshVisibility(true);
+            }
+
+            SolverHandler.UpdateSolvers = true;
+
+            IsBeingPlaced = true;
+        }
+
+        private void StopPlacement()
+        {
+            // Change the game object layer back to default which enables raycast hits again
+            gameObject.layer = DefaultLayer;
```

Comment:
Either two options here

1) Capture what the original layer was in StartPlacement and then re-apply here. Default isn't necessarily correct

2) Don't modify the gameobject layer at all. Just fire a warning message

Reply:
I am going to go with option 1, the layer change is required if the object is in the placing state.  If the game object layer is not changed to IgnoreRaycast while in the placing state, the head gaze ray will return a hit on the object itself instead of a surface beyond the object.